### PR TITLE
KEYCLOAK-6638 Support for adding audiences to tokens

### PR DESCRIPTION
**Pull request by mposolda**

### CODE REVIEW

**Comment by pdrozd** [comment]
@keycloak-ci-bot test

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/AudienceTest.java
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/AudienceTest.java]: Is it possible to change this behavior somehow and avoid having the frontend in the list of audiences ? Asking because people may want to have in access tokens only the audiences representing the resource servers where token will be used.

File: core/src/main/java/org/keycloak/representations/JsonWebToken.java
Comment (mposolda) [reply core/src/main/java/org/keycloak/representations/JsonWebToken.java]: Yes, Set is better. 

But it will require to change signature of JsonWebToken.getAudience/setAudience methods, so kept the String[] in the signatures for backwards compatibility (people may already use JsonWebToken directly in their application). Also the "aud" can be single String or an Array per JSON and hence we use this custom "StringOrArraySerializer" / "StringOrArrayDeserializer", which would probably need to be rewritten or different serializer used etc.

So the easiest would be to handle duplicates in the "addAudience" method. If you prefer to change it to Set, I can take a look at it (will require some work with the Serializer. For backwards compatibility it is probably ok if we just document in the migration guide). WDYT?

Comment (pedroigor) [reply core/src/main/java/org/keycloak/representations/JsonWebToken.java]: I think that handle duplicates is enough. Didn't consider backward compatibility, good point.

Comment (mposolda) [reply core/src/main/java/org/keycloak/representations/JsonWebToken.java]: @pedroigor I've updated PR to handle duplicates.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
LGTM, not sure if I did right by approving the PR as we are still waiting for @stianst feedback :)

**Comment by mposolda** [comment]
@pedroigor Thanks! I am not sure as well. But I am not going to merge anything without the ack from @stianst :)

**Pull request merged**
